adapter/pkg/messaging/azure_listener.go

```diff
@@ -19,126 +19,49 @@
-
-	var topicExistForFurtherProcess bool
-	var subscriptionCreatedForFurtherProcess bool
-	parentContext := context.Background()
-	logger.LoggerMgw.Info("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Starting broker consumer for topic name : " +
-		topicName)
-
-	uniqueID := uuid.New()
-	//In Azure service bus subscription names can contain letters, numbers, periods (.), hyphens (-), and
-	// underscores (_), up to 50 characters. Subscription names are also case-insensitive.
-	var subscriptionName = componentName + "_" + uniqueID.String() + "_sub"
-
-	// TODO: (dnwick) Handle retry logic in error situations
-
-	if !isTopicExist(topicName, availableTopicList) {
-		//create the topic
-		topicManager := ns.NewTopicManager()
-		ctx, cancel := context.WithCancel(parentContext)
-		defer cancel()
-
-		_, err := topicManager.Put(ctx, topicName)
-		if err != nil {
-			logger.LoggerMgw.Error("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Error occurred while trying to create " +
-				"topic " + topicName + " from azure service bus :%v", err)
-		} else {
-			logger.LoggerMgw.Info("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Topic " + topicName + "created")
-			topicExistForFurtherProcess = true
-		}
-	} else {
-		topicExistForFurtherProcess = true
+	dataChannel := make(chan []byte)
+	if strings.EqualFold(topicName, notification) {
+		dataChannel = AzureNotificationChannel
+	} else if strings.EqualFold(topicName, tokenRevocation) {
+		dataChannel = AzureRevokedTokenChannel
 	}
+	parentContext := context.Background()
 
-	if topicExistForFurtherProcess {
-		logger.LoggerMgw.Info("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Topic " + topicName + " exist. Can proceed")
-		subManager, err := ns.NewSubscriptionManager(topicName)
-
+	for {
+		//topic subscription client creation
+		topicSubscriptionClient, err := subscriptionMetaData.subscriptionManager.Topic.NewSubscription(subscriptionName)
 		if err != nil {
-			logger.LoggerMgw.Error("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Error occurred while trying get subscription" +
-				" manager from azure service bus for topic name " + topicName + ":%v", err)
+			logger.LoggerMgw.Errorf("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Error occurred while trying to create "+
+				"topic subscription client for %s from azure service bus for topic name %s:%v.",
+				subscriptionName, topicName, err)
 		}
-
-		//We are creating a unique subscription for each adapter starts. Unused subscriptions will be deleted after
-		// idle for three days
-		ctx, cancel := context.WithCancel(parentContext)
-		defer cancel()
-		_, err = subManager.Put(ctx, subscriptionName, opts...)
-
+		logger.LoggerMgw.Info("[TEST][FEATURE_FLAG_REPLACE_EVENT_HUB] Starting to receive messages for " +
+			"subscriptionName  " + subscriptionName + " from azure service bus for topic name " + topicName)
+		func() {
+			ctx, cancel := context.WithCancel(parentContext)
+			defer cancel()
+			err = topicSubscriptionClient.Receive(ctx, servicebus.HandlerFunc(func(ctx context.Context,
```

Comment:
Since method has the parent context and the code is inline,
I guess we could avoid func annotation

Reply:
no this is in a for loop so since we use defer cancel() it will cancel when the for loop finishers which is never as we are retrying infinitely. So having anonymous func will make sure context is within that scope and when it goes out of it because of an error the context will also canceled else there will be a memory leak